### ISSUE TITLE
ERR_SSL_VERSION_OR_CIPHER_MISMATCH on port 10000

### ISSUE THREAD

**Issue by African-Studios** [webmin/webmin#1854]
Recently renewed my ssl certificate with Let's Encrypt. Now miniserver.conf is producing ERR_SSL_VERSION_OR_CIPHER_MISMATCH on all browsers.  Websites that I am hostihg work find with httpd, but miniserver.conf is only thing on port 10000.

here is my conf 

port=10000
addtype_cgi=internal/cgi
realm=Webmin Server
logfile=/var/webmin/miniserv.log
errorlog=/var/webmin/miniserv.error
pidfile=/var/webmin/miniserv.pid
logtime=168
ssl=1
no_ssl2=1
no_ssl3=1
no_tls1=1
no_tls1_1=1
no_tls1_2=0
ssl_honorcipherorder=1
no_sslcompression=1
env_WEBMIN_CONFIG=/etc/webmin
env_WEBMIN_VAR=/var/webmin
atboot=1
logout=/etc/webmin/logout-flag
listen=10000
denyfile=\.pl$
log=1
blockhost_failures=5
blockhost_time=60
syslog=1
ipv6=1
session=1
premodules=WebminCore
userfile=/etc/webmin/miniserv.users
keyfile=/etc/letsencrypt/live/XXXXXXX/privkey.pem
passwd_file=/etc/shadow
passwd_uindex=0
passwd_pindex=1
passwd_cindex=2
passwd_mindex=4
passwd_mode=0
preroot=authentic-theme
passdelay=1
login_script=/etc/webmin/login.pl
logout_script=/etc/webmin/logout.pl
cipher_list_def=1
failed_script=/etc/webmin/failed.pl
error_handler_403=403.cgi
error_handler_404=404.cgi
error_handler_401=401.cgi
nolog=\/stats\.cgi\?xhr\-stats\=general
certfile=/etc/letsencrypt/live/XXXXXXX/cert.pem
ssl_redirect=0
no_resolv_myname=0
sockets=
extracas=/etc/letsencrypt/live/XXXXXXX/chain.pem
root=/usr/libexec/webmin
mimetypes=/usr/libexec/webmin/mime.types
server=MiniServ/2.013
ssl_hsts=0
ssl_cipher_list=ssl_cipher_list=ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!SRP:!CAMELLIA

**Comment by chris001** [comment]
Try changing this:
`ssl_cipher_list=ssl_cipher_list=`

to this:
`ssl_cipher_list=`

**Comment by African-Studios** [comment]
Things for the catch, Fresh eyes alwayse help. unfortunately that wasnt the issue.  I will try a catch all certificate next.

**Comment by jcameron** [comment]
Which Linux distribution and version are you running there? It's possible that your OpenSSL version isn't compatible with the new SSL key..

**Comment by African-Studios** [comment]
CentOS Linux 7.
Thanks for the feedback and no. SSL works on all domains but not the webmin
port 10000. Worked for years until I recently updated the certificate which
leads to believe that something changed in how miniserv works. Everything
else has been constant.

Babatunde Odesanya
Producer | Director | Writer
View my IMDB profile <Http://www.imdb.com/name/nm3547888/>
Follow me on Facebook

*African Studios*
Create and Deliver
347 688 7592
***@***.***

Proud member of Black and Latino Filmmaker's Coalition
visit BlackandLatinoFilm.com

**Comment by iliajie** [comment]
It's most probably SSL key is now using ECC algorithm by default.

**Comment by jcameron** [comment]
@African-Studios how did you request this cert?

**Comment by African-Studios** [comment]
Let's Encrypt  via webmin panel.

**Comment by jcameron** [comment]
ECC certs should work fine with Webmin already - at least they do in my tests.

Are you sure that the files in the cert= and key= lines in /etc/webmin/miniserv.conf have the same timestamp, indicating that they were issued together?

**Comment by cagivacode** [comment]
I just had this issue today...upgrading to 2.020...have been using ecdsa for six years(?).
I think it was the private key that was in ecdsa format. 
Could not get into webmin (auto forward to https).....the UI kept try to login by itself and set off fail2ban.
Generally unpleasant upgrade

**Comment by African-Studios** [comment]
@jcameron Yes, 

@cagivacode
 ssh 
cd  /etc/webmin
 nano miniserv.conf  
change ssl=1 to ssl =0 
systemctl restart webmin

then you will be able to log in using http:// and fix your config

**Comment by cagivacode** [comment]
yes that is how I got of the situation...and then I converted the SSL key to RSA. Turned SSL back on.

My point is that everything worked fine as was (includding SSL key) until 2.020
acme.sh defaults to ecdsa and I use it to supply wildcard Certs for the server.

**Comment by jcameron** [comment]
Does anything get logged to `/var/webmin/miniserv.error` when you start Webmin with an EC cert, or when you try to connect?

**Comment by cagivacode** [comment]
Not in the error log but in the regular log there is gibberish that started at the time of upgrade and stopped when I went to a RSA key.

400] "POST /session_login.cgi HTTP/1.1" 500 1652 "http://sebring12.xxx.com:5001/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36 Edg/113.0.0.0"
pool-108-16-220-xxx.phlapa.fios.verizon.net - - [15/Mar/2023:19:50:49 -0400] "�]</U|F�H{����B~ ��k�B��j(ň�� �4�韘��F=�\X���*�B�����;��) ���+�/�,�0̨̩����/5���#h2http/1.1-+��3+)" 400 1454 "" ""
pool-108-16-220-xxx.phlapa.fios.verizon.net - - [15/Mar/2023:19:50:49 -0400] "�ܩ" 400 1454 "" ""
pool-108-16-220-xxx.phlapa.fios.verizon.net - - [15/Mar/2023:19:55:39 -0400] "��`B"����/Rc�����u>b�Ϙ�	�ܓ�q a1�<�L�pX;����x�֦�m�Z��ot��� jj�+�/�,�0̨̩����/5���h2http/1.1#3+)JJ �KpTG�C�9���q�"�ԩqhE��>)T6]��P4�+**" 400 1454 "" ""
pool-108-16-220-xxx.phlapa.fios.verizon.net - - [15/Mar/2023:19:55:39 -0400] "����Y� Ӏ���kU�H8��13���@tW�0X ����](�(����q�hN�L@e��n%@��� ���+�/�,�0̨̩����/5�" 400 1454 "" ""
pool-108-16-220-xxx.phlapa.fios.verizon.net - - [15/Mar/2023:19:56:00 -0400]

**Comment by cagivacode** [comment]
I did not notice it was @jcameron that asked....

Detailed background for the server that had an issue:
That server runs webmin for general use and a panel for management of websites (not virtualmin). webmin has been kept up-to-date at all times.

The certificates are issued and installed directly by acme.sh...they were not installed or issued through webmin. 
Acme.sh was run without and key parameters so it was using default settings...although I believe that the certs have been ecdsa format for a long time. Originally issued by the other panel / certbot for a specific client reason, then updated regularly by acme.sh for the past few years.

I believe that both certbot and acme.sh now default to ecdsa and I think webmin defaulted to RSA in the 2.020 update. 

On my test system which is webmin/virtualmin the update went fine, but all the certs are issued/handled by them.

cheers,
red

**Comment by jcameron** [comment]
The issue may be that the particular kind of cert issued by acme.sh isn't what Webmin expects.

What's the exact command line you are running?

**Comment by cagivacode** [comment]
completely vanilla:

acme.sh --issue --dns dns_pdns --dnssleep 30 -f -d xxx.com -d *.xxx.com \
 --cert-file /etc/letsencrypt/live/xxx.com/cert.pem \
 --key-file /etc/letsencrypt/live/xxx.com/privkey.pem \
 --ca-file /etc/letsencrypt/live/xxx.com/ca.pem \
--fullchain-file /etc/letsencrypt/live/xxx.com/fullchain.pem

I have now added  " --keylength 4096"  which forces rsa

**Comment by jcameron** [comment]
Was the cert issued by Let's Encrypt? When I test this, it's always calling acme.zerossl.com

**Comment by cagivacode** [comment]
It is letsencrypt...I set it as default right when acme.sh switched to zerossl with...
/root/.acme.sh/acme.sh --set-default-ca  --server  letsencrypt

but it can be done with command line as well.

**Comment by jcameron** [comment]
Thanks - I requested a cert using that same command, and was able to install it into Virtualmin with no issues.

Which browser are you using in which you get that error message?

Also, does commenting out the `ssl_cipher_list=` line in `/etc/webmin/miniserv.conf` and then running `/etc/webmin/restart` help any?

**Comment by ajdonnison** [comment]
Had the same problem, with certbot added `--key-type rsa` and that solved it for me.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.

**Comment by iliaross** [comment]
Closing this as the reported failure appears to be caused by a stale/custom TLS cipher configuration rather than an active Webmin bug.
